### PR TITLE
Refactor JDocument so it can be autoloaded

### PR DESCRIPTION
**Pull request by wilsonge**
To test ensure that Joomla itself functions the same (just try a couple of pages). The error page works and any custom formats you might have (e.g. the VCF view for com_contact etc).

There should be no changes before and after the patch, however in the future you can now autoload properly in JDocument. Having things like the renderers now running autoloading is going to be important in the future for introduction of alternate output formats.

### CODE REVIEW

File: tests/unit/suites/libraries/joomla/document/json/JDocumentJSONTest.php [hunk 1]
```diff
@@ -7,16 +7,14 @@
...
 /**
- * Test class for JDocumentJSON
+ * Test class for JDocumentJson
  */
-class JDocumentJSONTest extends TestCase
+class JDocumentJsonTest extends TestCase
 {
 	/**
-	 * @var  JDocumentJSON
-\	 */
```
Comment (dgt41) [inline tests/unit/suites/libraries/joomla/document/json/JDocumentJSONTest.php]: oops

**Comment by izharaazmi** [comment]
:+1:  for adding support for `JDocumentRenderer[doc-type][rendered-type]`. How about now providing a standard set of `JDocumentJson` renderers? Should we?

**Comment by photodude** [comment]
I have tested this item :white_check_mark: successfully on aa7c947764fbfda79c875de803d9d5f730923598<br /><br />As per test instructions, tested that Joomla itself functions the same<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8905">issues.joomla.org/joomla-cms/8905</a>.</sub>

File: tests/unit/suites/libraries/joomla/document/json/JDocumentJSONTest.php (hunk 1 above)
Comment (wilsonge) [inline tests/unit/suites/libraries/joomla/document/json/JDocumentJSONTest.php]: It's a removed line. So all good now :)

**Comment by wilsonge** [comment]
@izharaazmi something like that won't be part of this PR. As long as people are extending the right classes (which we have checks for) then it's down to them to provide an implementation. In the future we could investigate some sort of interface. But it's likely going to be something JDocumentRenderer already supports anyhow

**Comment by izharaazmi** [comment]
Yeah, I understand that that can't be part of this PR. I was asking for the additional prospects. I think we can provide a full fledged structured rendering similar to that of html renderers, like response code, messages, body, redirects, callbacks, etc or whatever we can figure out when into it.

This can give us a well defined format for Joomla based JSON APIs as well. Same can be done for XML rendering for XML APIs.

I understand that this is not the part of Joomla project scope right now. But I wonder, what if?

Btw, JDocumentJSON right now appears to be just a barebone (sorry! I may be wrong here). It renders the output with `Content-Disposition', 'attachment'` anyway, which usually is not helpful always.
However, these can be discussed later in case we actually decide to work on this.

**Comment by wilsonge** [comment]
Exactly. There's definitely a lot of work to do with `JDocument`. I see this as the start of getting a lot of these issues fixed up

**Comment by mbabker** [comment]
JDocument and its subclasses aren't a proper request/response structure, nor even responsible for building the response.  It is at best a page rendering system; all of the response data is handled as part of the JApplicationWeb API.  JDocument does have a couple of calls into the application to set some response data based on the format, but that's the extent of what it's capable of at the moment.

**Comment by anibalsanchez** [comment]
I have tested this item :white_check_mark: successfully on aa7c947764fbfda79c875de803d9d5f730923598<br /><br />I have been working with this patch, it works OK.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8905">issues.joomla.org/joomla-cms/8905</a>.</sub>

**Pull request merged**